subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1087,6 +1093,29 @@ public void preventFromFurtherMutation() {
             configurationAttributes = new ImmutableAttributeContainerWithErrorMessage(delegatee, this.displayName);
             outgoing.preventFromFurtherMutation();
             canBeMutated = false;
+
+            if (isCanBeConsumed()) {
+                ensureUniqueAttributesOnConsumableConfigurations();
+            }
+        }
+    }
+
+    private void ensureUniqueAttributesOnConsumableConfigurations() {
+        if (configurationsProvider != null) {
+            final Set<? extends ConfigurationInternal> all = configurationsProvider.getAll();
+            if (all != null) {
+                final Set<? extends ConfigurationInternal> potentialDuplicates = configurationsProvider.getAll().stream()
+                    .filter(c -> c != this)
+                    .filter(c -> !c.isMutable())
+                    .filter(c -> c.isCanBeConsumed())
+                    .collect(Collectors.toSet());
+
+                for (ConfigurationInternal other : potentialDuplicates) {
+                    if (!other.getAttributes().isEmpty() && !getAttributes().isEmpty() && other.getAttributes().asMap().equals(getAttributes().asMap())) {
+                        throw new GradleException("Consumable configurations within a project must have unique attribute sets, but " + getDisplayName() + " and " + other.getDisplayName() + " contain identical attribute sets.");
+                    }
+                }
```

Comment:
This could be describing premature optimization, so feel free to punt on it if you think so.

If every configuration in a project ends up being made immutable, this is O(n<sup>2</sup>) in the number of configurations. It might be conceptually simpler to define this logic in the configuration provider, and then this implementation could trade-off memory for runtime by maintaining a hash map keyed by the attributes.

It is also a single location that can potentially deal with any synchronization too if that is needed.

Reply:
Doing a single pass through all the configurations in the provider would be preferable, and potentially faster, but I'm not sure it is possible.  We have to wait until after a configuration is locked against mutation to check it.  There doesn't seem to be any easy way to tell that all the configurations that will be locked in this way are finished, and no further locking will occur.  So because of this we will never know if we are completely done mutating unless all configurations happen to be locked against mutation - this likely won't happen most of the time.

While the algorithm's characteristics aren't great, in practice the number of configs which are actually included in the comparison now should be tiny (if not, that itself is probably a code smell in the project setup).  In any case, performance tests over representative projects will be run as part of CI builds.  If this change adds too much additional wait time, those checks will fail and we'll know it is unacceptable.

Reply:
Sounds good to me!